[quant] Enable per-channel quantization for LSTM Modules

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39666 [quant] Enable per-channel quantization for LSTM Modules**
* #39604 [quant] Add reduce_range params for quantized_lstm

Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingDynamic.test_per_channel_lstm_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21977601](https://our.internmc.facebook.com/intern/diff/D21977601)

supriyar committed:
[quant] Enable per-channel quantization for LSTM Modules
Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingDynamic.test_per_channel_lstm_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize.py:
```diff
@@ -866,6 +866,49 @@ def test_default_quantized_lstm(self):
+        hx = torch.tensor(h0_vals, dtype=torch.float).unsqueeze(0)
+        cx = torch.tensor(h0_vals, dtype=torch.float).unsqueeze(0)
+        hiddens = (hx, cx)
+        quant_out, quant_hidden = model_quantized(x)
```
We need to check for scriptability and serialization, like in other tests. My PR should add this support for the baseline LSTM case, but hasnt landed yet.
```
            def checkQuantized(model):
                self.assertEqual(type(model.fc1), torch.nn.Linear)
                self.checkDynamicQuantizedLinear(model.fc2, dtype=dtype)
                self.checkScriptable(model, self.calib_data, check_save_load=True)
```

supriyar committed:
Update on "[quant] Enable per-channel quantization for LSTM Modules"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingDynamic.test_per_channel_lstm_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Enable per-channel quantization for LSTM Modules"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingDynamic.test_per_channel_lstm_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Enable per-channel quantization for LSTM Modules"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingDynamic.test_per_channel_lstm_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
